spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+    1. [=Assert=] that |context| is not null.
+    1. If |key|'s [=string/length=] exceeds the [=key/maximum length=], reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |value|'s [=string/length=] exceeds the [=value/maximum length=], reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |options| is invalid, reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |options| is empty, create |options|.
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |result| be null.
+        1. If |options|["`ignoreIfPresent`"] is true:
+            1. Let |currentValue| be the result of running [=shared storage database/retrieve an entry from the database=] with |context| and |key|.
+            1. If |currentValue| is not null and |currentValue| is not undefined, resolve |promise| as undefined.
+            1. If |currentValue| is null, reject |promise|.
+            1. Otherwise, let |result| be the result of running [=shared storage database/store an entry in the database=] with |context|, |key|, and |value|.
+        1. Otherwise, let |result| be the result of running [=shared storage database/store an entry in the database=] with |context|, |key|, and |value|.
+        1. If |result| is false, reject |promise|.
+        1. Otherwise, resolve |promise| as undefined.
+    1. Return |promise|.
+  </div>
+
+  <div algorithm>
+    The <dfn method for="WindowSharedStorage">append(|key|, |value|)</dfn> method steps are:
+
+    1. Let |promise| be a new [=promise=].
+    1. Let |context| be {{WindowSharedStorage}}'s {{Window}}'s [=Window/browsing context=].
+    1. [=Assert=] that |context| is not null.
+    1. If |key|'s [=string/length=] exceeds the [=key/maximum length=], reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |value|'s [=string/length=] exceeds the [=value/maximum length=], reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |result| be null.
+        1. Let |currentValue| be the result of running [=shared storage database/retrieve an entry from the database=] with |context| and |key|.
+        1. If |currentValue| is not failure:
+            1. Let |list| be a new [=list=].
+            1. [=list/Append=] |currentValue| to |list|.
+            1. [=list/Append=] |value| to |list|.
+            1. Let |newValue| be the result of running [=string/concatenate=] on |list|.
+            1. Let |result| be the result of running [=shared storage database/store an entry in the database=] with |context|, |key|, and |newValue|.
+        1. Otherwise, let |result| be the result of running [=shared storage database/store an entry in the database=] with |context|, |key|, and |value|.
+        1. If |result| is false, reject |promise|.
```

Comment:
Promises should be rejected with values like TypeError.  Something like `reject |promise| with a {{TypeError}}`.  Here and elsewhere.